xls/examples/bitonic_sort.x

```diff
@@ -0,0 +1,220 @@
+////////////////////////////////////////////////////////////////////////////////
+// Bitonic sort
+// A template function for sort an array of u32.
+// The array size must be a power of two.
+// The template takes the log2 of the size of the array as parameter.
+//
+// For example, assuming "array" is of type u32[8].
+// Then, bitonic_sort<u32:3>(array) will return the sorted array.
+////////////////////////////////////////////////////////////////////////////////
+
+fn bitonic_sort<LOG_N: u32, N: u32 = { u32:1 << LOG_N }>(array: u32[N]) -> u32[N] {
+	let (result, _) = for (_, (array, k)) in u32:0..LOG_N {
+		let (result, _) = for (_, (array, j)) in u32:0..LOG_N {
+			if (j == u32:0) {
+				(array, j)
+			} else {
+				let result = for (i, array) in u32:0..N {
+					let l = i ^ j;
+					if (l <= i) {
+						(array)
+					} else {
+						if ( (((i & k) == u32:0) && (array[i] > array[l])) || (((i & k) != u32:0) && (array[i] < array[l])) ) {
+							let i_element = array[i];
+							let l_element = array[l];
+							update(update(array, i, l_element), l, i_element)
+						} else {
+							(array)
+						}
+					}
+				} (array);
+				(result, j / u32:2)
+			}
+		} ((array, k / u32:2));
+		(result, k * u32:2)
+	} ((array, u32:2));
+	result
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////////////
+
+#[test]
+fn bitonic_8_test() {
+	let _ = assert_eq(
+		u32[8]:[
+		223,237,289,466,673,686,707,916,
+		],
+		bitonic_sort<u32:3>(
+		u32[8]:[
+		707,223,686,289,237,673,466,916,
+		]));
+	()
+}
+
+#[test]
+fn bitonic_128_test() {
+	let _ = assert_eq(
```

Comment:
curious if we could `#[quickcheck]` this?
https://google.github.io/xls/dslx_reference/#quickcheck